_posts/2025-05-16-last-ten-years.markdown

```diff
@@ -0,0 +1,137 @@
+CHERIoT had to shrink this further and had other microarchitectural constraints for short pipelines, which is why we support two capability formats for different scales of devices.
+
+And, of course, the MIPS prototypes have been replaced with RISC-V.
+64-bit big-endian MIPS was never well supported and RISC-V reached parity quite quickly and provided a better ecosystem for experimentation.
+
+# CHERIoT adds some more things
+
+CHERIoT started in 2019 and tried to scale the ideas down.
+The 2015 paper didn't mention a 32-bit option at all.
+The [2018 CheriRTOS paper based on Hongyan Xia's PhD work](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201810-iccd2018-cheri-rtos.pdf) was, I think, the first public work to scale down to 32 bits.
+This had a lot of limitations from the encoding (Hongyan was one of the original CHERIoT team and we learned a lot from his experience).
+
+CHERIoT adds more kinds of sentries, temporal safety, a richer set of permissions, and closer co-design between the hardware and software.
+
+The original DARPA challenge asked what you would change 'if you could change anything'.
+In some ways, CHERI was the least ambitious project on that programme because we cared a lot about backwards compatibility and incremental adoption.
+The 2015 paper makes it clear that you can simply opt out of CHERI for some processes (you don't need to recompile the world) and that you can use legacy ABIs in sandboxes with a bit of glue code doing copying.
+This kind of incremental adoption story is really important, but on embedded systems incremental adoption looks quite different.
+Most embedded code has few dependencies from the host environment, which gave us a lot more freedom to build efficient models that scale right down to tiny devices.
+Many of thees would be possible for green-field software on bigger devices, but you can't build those unless CHERI is widely deployed and you can't deploy CHERI widely without CHERI-enabled CPUs running existing software stacks.
+
+# We aren't finished yet!
+
+The CHERI project has been going on for 15 years now, and CHERIoT for six of those.
+CHERI is a bigger change than adding memory management units (MMUs) to systems.
+The core ideas in MMUs came from the 1970s, yet people are still coming up with new variations on MMUs and new software abstractions that use them.
+We are now at the state with CHERI where we were with MMUs in the 1990s: We understand some *very* useful things that can be built with CHERI and we are now good at building those things, but there's still a large set of things that we haven't considered.
+
+I expect CHERI to enable a lot of new software patterns that will provide a much larger benefit than anything that we've seen so far.
+Fixing 70% of security vulnerabilities and providing fine-grained compartmentalisation for fearless code reuse is just the start!
```

Comment:
FWIW, I think this is a key idea -- simple (but strong) memory safety is just the start, because support for granular compartmentalisation is about enabling new software architecture.